### PR TITLE
Attempt to improve upload performance

### PR DESCRIPTION
**Pull request by webbnh**
The recent "Ops Review" session revealed that upload requests time out if the file approaches 7Gb in size.  Profiling the execution indicates that the bulk of the time is spent extracting the `metadata.log` file from the results tarball.  This is due to a "feature" of tar archives:  the same file may appear more than once, and so an extraction has to search the entire archive to ensure that the member being extracted is the most up-to-date (i.e., last) one, and this is punishing as the archive size grows.  (Also, the Python Standard library `tarfile` package is not as performant as the standalone `tar(1)` utility, and this costs additional time.)

This PR changes the Server to spawn a subprocess to run the `tar(1)` utility, which allows us to specify the `--occurrence` option which allows us to abort the search after the first instance of the member is found.  Since Pbench owns the construction of the archive, we know that each member occurs only once, so there is no need to search for other instances of it.  Also, since the `metadata.log` file occurs near the beginning of the archive, by aborting the search when we find the first instance of the member, we are able to reduce the overall search time from multiple minutes to a fraction of a second.

This PR consists of a series of commits.  Two of them are the changes to the cache manager module to add the subprocess execution and the changes to the corresponding unit tests.  Two of them are passes over the cache manager module and the corresponding tests to "pick lint".  The others are iteration on the respective changes to the code and the tests.

### CODE REVIEW

File: lib/pbench/server/cache_manager.py
Comment (dbutenhof) [inline lib/pbench/server/cache_manager.py:510]: I'd debated this change earlier, and decided to keep the str-ification here and let callers use `Path` "math" without having to remember to convert. And, technically, (and despite the simplistic type hint) that made this a "PathLikeObject" (a type that doesn't actually exist and is awkward to construct), which I rather liked. I'm not sure how I feel about the change...

File: lib/pbench/server/cache_manager.py
Comment (dbutenhof) [inline lib/pbench/server/cache_manager.py:214]: Which is interesting; as it's called deep within werkzeug after we've returned the `Response` object, I have no idea what this will do... but it's likely a 500 error. Would it be better to catch that (maybe with a log) and avoid propagating the error?

File: lib/pbench/server/cache_manager.py
Comment (dbutenhof) [inline lib/pbench/server/cache_manager.py:582]: Yeah, too bad `tar` doesn't have useful (or even well defined) non-zero return codes. (Sigh.)

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/server/cache_manager.py
Comment (webbnh) [reply lib/pbench/server/cache_manager.py:214]: To my way of thinking, if this exception gets raised, it's pretty much a panic situation:  we've sent a `SIGKILL` to the subprocess, we've drained the pipes, and it still hasn't gone away after 60 seconds...as far as I can see, there's really nothing left to be done except cry for Mommy....

File: lib/pbench/server/cache_manager.py
Comment (webbnh) [reply lib/pbench/server/cache_manager.py:582]: Actually, the error text (at least the ones that we're interested in) come from `strerror()` which, by this point, is pretty stable and standard...so we _could_ consider recognizing the messages...but I'm satisfied with the current arrangement.

**Comment by webbnh** [comment]
This is now ready for review.

File: lib/pbench/server/cache_manager.py
Comment (dbutenhof) [inline lib/pbench/server/cache_manager.py:592]: If we successfully create the subprocess, but fail out before returning an `Inventory` (e.g., on a timeout), we should probably make an attempt to shut things down as gracefully as possible? Basically the `Inventory.close`, perhaps?

File: lib/pbench/server/cache_manager.py
Comment (dbutenhof) [inline lib/pbench/server/cache_manager.py:1171]: 👍🏻

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_cache_manager.py:947]: Wow. That's a lot.

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_cache_manager.py:1044]: Hmm.

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_cache_manager.py:1061]: It seems what we're missing here is an explicit confirmation that we *expected* a `CacheExtractBadPath`, not just that we expected a tar path and for `Popen` to not fail. That is, that `CacheExtractBadPath` wasn't raised erroneously...

**Review by dbutenhof — CHANGES_REQUESTED** [review]
Mostly great, but unless I'm missing something, a few tests seem a bit weak?

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_cache_manager.py:532]: Not at all sure how I feel about that, but I'll let it go. 😆

**Review by dbutenhof — COMMENTED** [review]
`test-cache-manager.py` needs to be blackened

**Review by dbutenhof — APPROVED** [review]
`test_cache_manager.py` needs to be blackened... but I'll grant a symbolic victory lap ...

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_cache_manager.py:532]: Yeah...I have all sorts of conflicting desires on this one.

I would like to have 100% coverage of the CUT, with 100% execution of the unit test code.  This suggests that I should just remove the unused mock.

On the other hand, I know that the CUT _would_ call `resolve()` if execution happened to follow the other branch of a conditional in the function, so there is a real danger of calling it, which suggests that we should have a rigorous mock.

On the gripping hand, I'm not sure how much it would really matter if the real `resolve()` got called, and the results of that might be just as obvious as raising the assertion exception.

I don't really want the extra complication in the unit test, especially when the key part of the code won't even execute because we never call the mock....

🦐

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_cache_manager.py]: Well, now...since I added a call to `kill()` in `Tarball.extract()` for the creation-timeout, now this has to be a "real mock"...so, let's have another round!

**Review by ndokos — APPROVED** [review]
Still looking at the test code, but the rest LGTM.

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_cache_manager.py:1043]: I suppose that's fine that you're mocking `Inventory.close` to a different `close` you also don't expect to be called, and you cover all of the `Inventory` class below, but it looks odd and the self-documentation here (plus the assert if it *is* called) could be better. Ah, well.

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/test/unit/server/test_cache_manager.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_cache_manager.py:1043]: It _should_ be fine to mock `Inventory.close()` with a different `close()`, given that they are all supposed to be providing the same `io.IOBase.close()` interface!  😆  (And, yeah, since I expect none of them to be called, using separate implementations seems silly and redundant.  😉)

Nevertheless, I don't disagree that, on the surface, it looks funny.  🙂

As for the documentation...well, Nick is still looking at the test code, so there is still a chance to fix that.  🙃  What in particular would you like to see?

Comment (webbnh) [reply lib/pbench/test/unit/server/test_cache_manager.py:1043]: > once you start to write that and think about how it'll look, you'll probably just write the "proper" mock instead.

One of the problems that I'm trying to address is to minimize the amount of code in the unit tests which is **not** executed per the coverage report.  So, adding an additional mock which causes the test to fail if it is called actually moves away from that goal.

> the clearest documentation would be a mocked `Inventory.close` that doesn't claim to be a stream close.

I find this to be curious, because, AIUI, we want `Inventory` to act like a stream, right?  (I mean, it provides all the same methods with basically the same semantics.)  So, it makes sense to me to use a stream-close mock for `Inventory.close()` (in this weird instance, anyway)....

>  Second best would be a comment that we're just being lazy here because it doesn't really matter as we don't expect it to be called.

I'll do that, if you think it's worth it (although, I don't think I'll term it as "lazy", because that's not really what it's accomplishing for me... 🙂).

Comment (ndokos) [reply lib/pbench/test/unit/server/test_cache_manager.py:1043]: You should not wait for me to check the test code before you merge. I don't know how long it will take me :grin:

Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_cache_manager.py:1043]: > I find this to be curious, because, AIUI, we want `Inventory` to act like a stream, right? (I mean, it provides all the same methods with basically the same semantics.) So, it makes sense to me to use a stream-close mock for `Inventory.close()` (in this weird instance, anyway)....

`Inventory` *composes* a stream and a "stream provider" so they can be managed together. That composition is critical, and eliding that to mock it as if it *were* a stream just seems really obscure and awkward to me. But what you've got will work, even if it might confuse a future reader who even notices it. I approved, so I'll leave the rest to you.

**Pull request merged**
